### ISSUE TITLE
Module/Idea Suggestion List.

### ISSUE THREAD

**Issue by Xeddius** [GitSquared/edex-ui#481]
**Perceived idea awesomeness**
<details><summary>Wow, you don't have this yet???</summary>
Web Browser Module,
Background Module,
Media Module/s,
Security/Lock Module.
</details>
<details><summary>Would be super cool.</summary>
Temperature Module,
Volume Control Module.
VNC Module,
Custom Text/Matrix Module.
</details>
<details><summary>I'm not sure a lot of people would use it, but I'd love to have it.</summary>
Self Destruct/Panic button/Red Alert Module.
</details>
<details><summary>Kind of a gadget, honestly, but why not?</summary>
Rainbow/Party mode Module.
</details>

**Perceived idea difficulty**
- [0] Quit your job first.
- [3] Would need some work, but nothing impossible.
- [7] Shouldn't be too hard to make.
- [1] Like, *I* could do it.

---
**Simple/-ish ideas:**
- Web Browser Module. (Requires some work, but still might be possible.)
- Background Module. (changes background to image, slideshow, or video/animation.)
- Media Module/s. (Show image/s, video, slideshows, music/audio, etc.)
- ~~Volume Control Module.~~ (Not feasable)
- Rainbow/Party mode. (Changes colors of theme over time. Not super useful, just really fun!)
- Security/Lock Module. (Lock Screen, alarm for wrong password, eDEX-UI only, non-system lock.)
- "[noVNC](https://github.com/novnc/noVNC)" module for connecting to another device. (a phone or another computer.) *(novnc is based on javascript and html, porting _should_ be pretty simple.)
- Text Module. (Print lines from a text file, Monitor a log file, or be set to rain code like in the matrix, should be pretty easy to make as the startup sequence already prints scrolling text.)
---
**Complex ideas/Stretch Goals:**
- LCARS Theme/Module. (Not likely to be super useful, but it'd be really neat, I'll likely attempt this in the future.)
---
**In Progress/Planned Modules:**
- [x] Space/Memory Module.
- [x] External Module Support.
- [x] Temperature Module.
- [x] Disk Activity Module.
---
**Completed Modules:**
- *placeholder*
---
**Mockup/Screenshots/Concepts**

Browser Module Concept: ![edex-main-browser-interface](https://user-images.githubusercontent.com/6382895/56373617-598c4400-61cf-11e9-9906-6e3eabab23a5.png)
Text Module Concept: ![animated](https://user-images.githubusercontent.com/6382895/56390410-f8c63100-61f9-11e9-8f80-22a182ff3f23.gif)

---

(UPDATE!)
Someone created a video player for Electron using Node.js, I'm sure it can be useful for helping to create a media module in edex-ui! It's called [Playback](https://github.com/mafintosh/playback)

This app could become a pretty practical UI replacement, with some of these suggested modules.
 I would definitely love to run this as the primary GUI on a minimal Arch Linux install! (probably running in i3 or another simple/lightweight window manager.)
I'm working on learning Javascript. I might be able to make one or a few of these modules sometime in the future. I will update this post with more ideas and possibly submit code as I continue to research and learn more about Electron + Javascript.

**Comment by GitSquared** [comment]
Disk activity is something that I'm ready to implement on *nix systems, but I'm waiting on some compatibility improvements upstream so it can also work reliably on Windows. I fully agree that this is a must-have missing feature!

Regarding disk space, there's already the mountpoint usage indicator below the filesystem display. Do you think it would be nice to also have an indicator for total physical drive usage?

**Comment by Xeddius** [comment]
That would indeed be nice. Would be cool to have an icon with either a circle, line, or chart next to it. A setting to specify a mount-point would be interesting as well so you can see how much space is on flash-drives, etc.

**Comment by GitSquared** [comment]
I'll look into it. Regarding mount-point usage, the indicator follows the mountpoint of the current CWD, so you can `cd` into your flash drive to see how much space is remaining ;).

**Comment by GitSquared** [comment]
@Xeddius Wow, you added a lot of things here since the last time I read this thread.

Here's some feedback:
- Web Browser module: While we do have Chromium's web engine at our disposal, implementing an actual web browser is a ton of work, a ton of code, and lots of things can go wrong. I recommend using CLI web browsers like `links` or `browsh`, and if you want Netflix, open up a real full-featured GUI browser.
- Background module: Nothing undoable here, but could be kind of bloated for the main eDEX release itself. Maybe as an addon once #334's finally out?
- Media module: We could leverage Chromium's web engine and the existing tab system to display videos/images/docs opened through the filesystem module, I like this idea.
- Temperature module: CPU temp is already displayed on *nix systems. Other sensors are currently not supported by our back-end `systeminformation` library, so I'd need to either find an additional NPM module that does this or send some patches upstream.
- Volume control: Not possible. eDEX doesn't run with enough privileges on any platform to change system-wide parameters like this - also, eDEX is not a WM/DE so I won't be looking to implement those kind of things.
- Rainbow/Party mode: Sounds fun, but as with the Background module, should probably be some external addon.
- Lock module: See "Volume control" above; eDEX is not a WM/DE and making it emulate real higher-level system features is both hard and dangerous.
- "noVNC" mode: There is [ongoing work](https://github.com/GitSquared/edex-ui-server) (progressing slowly, I'll admit) to create a server-side component that would communicate with a "client" eDEX interface, allowing remote monitoring and control without running a user interface on servers.
- Matrix mode: Add that to the list of "extras". Or perhaps an April Fools special for next year, who knows.

Thanks for sharing all the great ideas you're having - also, your concept screenshots are truly inspiring, I love 'em. Keep those coming!

**Comment by GitSquared** [comment]
Also, regarding this:

> I would definitely love to run this as the primary GUI on a minimal Arch Linux install! (probably running in i3 or another simple/lightweight window manager.)

I've had some dude who bought a 22" touchscreen and a tiny Intel NUC reach out to me over Discord to make a custom, ultra-lightweight Debian install that would just run eDEX at startup, directly over X.org. We struggled a bit but I have some scripts and workarounds if you ever want to attempt this kind of thing.
On a "real" computer however, you'll indeed probably want to have a WM like i3 so you can run other apps.

**Comment by Xeddius** [comment]
^_^ I would most definitely like the scripts and workarounds. (might be able to do something with a Raspberry pi 3 or Rock64). Would it be possible to make a lock for eDEX-UI itself? (instead of a system-wide lock). A text module could tail log files or display a continuously scrolling stream of text from a file for the scifi/hacker aesthetic. 
![animated](https://user-images.githubusercontent.com/6382895/56390410-f8c63100-61f9-11e9-8f80-22a182ff3f23.gif)

**Comment by GitSquared** [comment]
@Xeddius [Here's a gist](https://gist.github.com/GitSquared/036678f0281a38ddd13e5df70cd0276b) with a little autostart script for eDEX on an almost-blank Debian install, running directly on X. I'm not sure if you can directly use this yourself, but It might be a good starting point.

Making a lock for eDEX itself would be doable, but I'm seriously questioning the usefulness of such a feature - notably when you could just pop up a dev tool window and remove the lock.

Regarding the background, I'm not sure I understand fully what your question is - the grid dot pattern behind the UI is just made with CSS gradient, and anything you can create in CSS - including images using blob URLs - can be used as a background.


Now, onto the fun things - during the early, pre-release days of eDEX, I actually implemented a system log module: 9b55675f50d1075e21e159f5ab63299d9c386320
This would start a `dmesg` child process and make it output to a temporary log file that the module would watch and mirror on the UI.

It worked really well, except that running `dmesg` required that eDEX asks for root privileges - something that was achieved using `sudo-prompt` - which, well, kind of sucked. Also, of course, this was a *nix-only feature and it made the gap between the Windows and Linux/macOS versions even larger, so I reverted it and it never got released.

That said, if we figure out some sufficiently interesting user-accessible system log that we can retrieve live on each OS, well, it would be trivial to update the old code and implement that back into eDEX.

**Comment by Xeddius** [comment]
The lock can also double as a screensaver kind of feature. It's mainly cosmetic, and the average person who doesn't know anything about edex likely wouldn't think to open a developer console. (It's mainly just pretty and adds to the immersion.)

Just wanted to know how the background works so I could change the colors of the dots and make different backgrounds.

I'm able to run dmesg without root on Linux Mint.
As for a dmesg equivalent for Windows there is
`wevutil qe system /f:Text /rd | more`

I figure it'd be nice to have a configurable log module that can monitor even text files, like if you wanted to monitor the progress of a specific program.

**Comment by Xeddius** [comment]
When changing colors of the theme the default colors of --color_grey never change with different themes and always stays a slightly blue color.

**Comment by GitSquared** [comment]
`dmesg` is root locked for me on Debian:
![image](https://user-images.githubusercontent.com/24496417/56886280-eb693c00-6a6e-11e9-920e-a43716cca101.png)

I'll look into it. Thanks for the Windows command - the idea to be able to watch any text file sounds promising.

That said, like most ideas you proposed in this thread, I feel like we should wait for the addon system to be implemented. Plus, launching that feature with already a small library of interesting optional modules would be awesome.
(I'm still planning on implementing the disk activity thing into the base eDEX build, I'm just holding on for Windows support).

I know it can be frustrating to get that kind of answer but rest assured that we're getting closer to an addon system implementation everyday. As you might have seen on the #334 thread, I'm waiting for remote monitoring first so I can expose a stable, mature API to addons that will automatically pipe through any eventual remote server.

I should be studying for exams but I kind of got a breakthrough on the technical side with the remote feature and I'm hyped as shit so, fingers crossed I'll get all this done in a reasonable time frame.

---

As for the dots grid background, here's the code that's used:
https://github.com/GitSquared/edex-ui/blob/04613fff5116ded9f27d350976cad4f41ab7027a/src/assets/css/main.css#L17

`--color_light_black` is the background color that's used to define the background and also in a lot of modules that need to recreate parts of the background - like the terminal in non-transparent mode.
`--color_grey` is actually the color of the dots, and is only used for that; setting it to `#ff00000` produces something like this:

![image](https://user-images.githubusercontent.com/24496417/56888062-70565480-6a73-11e9-9673-852499364c79.png)

Most themes don't play around with that color indeed - I believe only the `red` and `interstellar` ones do so.

Those CSS variables names date back to the infancy of this software and they're confusing at best, I should probably refactor this.

**Comment by GitSquared** [comment]
(note that this is properly documented in the [Wiki](https://github.com/GitSquared/edex-ui/wiki/Themes), though.)

**Comment by Xeddius** [comment]
I had an interesting Concept/Idea the other day, what about Shaders/CRT emulation? Would this at all be possible? I've simulated an example of what this feature might look like, below.
![Screenshot from 2019-05-04 17-39-11](https://user-images.githubusercontent.com/6382895/57185120-b9ebd880-6e93-11e9-8b75-b2ac5ed15b6b.png)

**Comment by GitSquared** [comment]
@Xeddius I like this idea, reminds me of `cool-retro-term`. I'm looking around a bit but can't figure out how that would be possible on a web interface. I think CSS magic could make something interesting though, I might play around with the idea next weekend.

**Comment by masivi** [comment]
The web browser module is a great idea! It would be brilliant !!!!

**Comment by Reshiram110** [comment]
I love the file viewing/matrix idea. If I could view text files (and possibly even edit them) in edex, it would be an amazing feature.

**Comment by GitSquared** [comment]
@Reshiram110 That wouldn't be too hard to implement, but while I see the usefulness of previewing images, I'm wondering why you would be so psyched about editing text files...?
I mean, it's a task that is very well handled by a lot of terminal-based solutions (vim, nano, macro, etc).

Honest question.

**Comment by Reshiram110** [comment]
@GitSquared Valid point, I wasn't thinking of terminal editors when I made the comment.

However, I would like to note that you cannot edit files currently when you click on them in the file explorer.
Adding in a custom editor module would allow us to say click a .log file and edit it in the custom editor UI.

**Comment by computamike** [comment]
How about a slack client?  Feel like I want to live in my Tron Desktop, and that includes sending slacks to my buddies.  Might have a go at it.. I mean... How hard can it be right?

**Comment by Xeddius** [comment]
@computamike Look into "https://github.com/erroneousboat/slack-term" it's a slack client for the terminal, something like that should work fine.
